### PR TITLE
composition: implement @composeDirective

### PR DESCRIPTION
**Pull request by tomhoule**
This is more permissive than the spec: it will compose directives without definition and without @link imports. It merges instances of the directive with the same arguments across subgraphs.

closes GB-5399

### CODE REVIEW

File: engine/crates/composition/src/ingest_subgraph/directives.rs [hunk 1]
```diff
@@ -109,3 +125,358 @@ pub(super) fn ingest_keys(
...
+            ));
+        }
+    }
+
+    directive_matcher.composed_directives = composed_directives;
+
+    directive_matcher
+}
+
+/// This struct is the source of truth for matching federation directives by name when ingesting a
```
Comment (tomhoule) [inline engine/crates/composition/src/ingest_subgraph/directives.rs:177]: From here, this file is 99% moved. Changes: removed FederationDirectivesMatcher to DirectiveMatcher and added the two struct fields about composed directives.

File: engine/crates/composition/tests/composition/composed_directives_basic/federated.graphql [hunk 2]
```diff
@@ -54,8 +54,8 @@ type ObserverDetails {
     observerType: ObserverType! @join__field(graph: OBSERVATIONS)
 }
 
-type BirdSighting {
-    bird: Bird! @join__field(graph: SIGHTINGS)
+type BirdSighting @deprecated(reason: "we haven't seen any birds in a while :(") {
```
Comment (tomhoule) [inline engine/crates/composition/tests/composition/composed_directives_basic/federated.graphql:57]: Looks like we were missing deprecated here. Thanks to the recent reorganization that is prolonged in this PR, handling of directives is uniform across locations and things like that are less likely to happen.

File: engine/crates/composition/src/compose/directives.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
+use super::*;
+
+pub(super) fn collect_composed_directives<'a>(
```
Comment (tomhoule) [inline engine/crates/composition/src/compose/directives.rs:3]: Logic gathered from different places and centralized.

File: engine/crates/composition/src/compose/directives.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                    )
+                })
+                .map(|arg| vec![arg])
+                .unwrap_or_default(),
+        });
+    }
+
+    for container in containers {
+        tags.extend(container.tags().map(|t| t.id));
+        is_inaccessible |= container.inaccessible();
```
Comment (yoav-lavi) [inline engine/crates/composition/src/compose/directives.rs:30]: Might be a bit cleaner to do this separately or spell it out more, was a bit surprised to see this

File: engine/crates/composition/src/compose/directives.rs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            name: ctx.insert_static_str("tag"),
+            arguments: vec![(ctx.insert_static_str("name"), federated::Value::String(name))],
+        });
+    }
+
+    composed_directives.extend(extra_directives);
+    composed_directives
+}
+
+fn subgraphs_value_to_federated_value(value: &subgraphs::Value, ctx: &mut ComposeContext<'_>) -> federated::Value {
```
Comment (yoav-lavi) [inline engine/crates/composition/src/compose/directives.rs:62]: Personally I'd expand single letter variables

**Review by yoav-lavi — COMMENTED** [review]
Nice work! See comments

File: engine/crates/composition/src/compose/directives.rs (hunk 5 above)
Comment (tomhoule) [reply engine/crates/composition/src/compose/directives.rs:62]: good point, it was a bit lazy, will do

File: engine/crates/composition/src/compose/directives.rs (hunk 4 above)
Comment (tomhoule) [reply engine/crates/composition/src/compose/directives.rs:30]: yep, I'll see how to make it more readable. At the very least with a comment.

File: engine/crates/composition/src/compose/directives.rs (hunk 5 above)
Comment (tomhoule) [reply engine/crates/composition/src/compose/directives.rs:64]: done

File: engine/crates/composition/src/compose/directives.rs (hunk 4 above)
Comment (tomhoule) [reply engine/crates/composition/src/compose/directives.rs]: done

**Review by yoav-lavi — APPROVED** [review]
Nice work!

**Pull request merged**